src/wp-includes/html-api/class-wp-html-template.php

```diff
@@ -0,0 +1,189 @@
+	 *    If provided any other type of value the attribute will be ignored and its existing value persists.
+	 *
+	 *  - If multiple HTML attributes are specified for a given tag they will be applied as if calling
+	 *    `set_attribute()` in the order they are specified in the template. This includes any attributes
+	 *    assigned through the attribute spread syntax.
+	 *
+	 *  - Substitutions in text nodes may only contain string values. If provided any other type of value
+	 *    the placeholder will be removed with nothing in its place.
+	 *
+	 *  - This function currently escapes all value provided in the arguments array. In the future
+	 *    it may provide the ability to nest pre-rendered HTML into the template, but this functionality
+	 *    is deferred for a future update.
+	 *
+	 *  - This function will not replace content inside of SCRIPT, or STYLE elements.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @access private
+	 *
+	 * @param string $template The HTML template.
+	 * @param string $args     Array of key/value pairs providing substitue values for the placeholders.
+	 * @return string The rendered HTML.
+	 */
+	public static function render( $template, $args = array() ) {
+		$processor = new self( $template );
+		while ( $processor->next_token() ) {
+			$type = $processor->get_token_type();
+			$text = $processor->get_modifiable_text();
+
+			// Replace placeholders that are found inside #text nodes.
+			if ( '#funky-comment' === $type && strlen( $text ) > 0 && '%' === $text[0] ) {
+				$name  = substr( $text, 1 );
+				$value = isset( $args[ $name ] ) && is_string( $args[ $name ] ) ? $args[ $name ] : null;
+				$processor->set_bookmark( 'here' );
+				$processor->lexical_updates[] = new WP_HTML_Text_Replacement(
+					$processor->bookmarks['here']->start,
+					$processor->bookmarks['here']->length,
+					null === $value ? '' : esc_html( $value )
+				);
+				continue;
+			}
+
+			// For every tag, scan the attributes to look for placeholders.
+			if ( '#tag' === $type ) {
+				foreach ( $processor->get_attribute_names_with_prefix( '' ) ?? array() as $attribute_name ) {
+					if ( str_starts_with( $attribute_name, '...' ) ) {
+						$spread_name = substr( $attribute_name, 3 );
+						if ( isset( $args[ $spread_name ] ) && is_array( $args[ $spread_name ] ) ) {
+							foreach ( $args[ $spread_name ] as $key => $value ) {
+								if ( true === $value || false === $value || null === $value || is_string( $value ) ) {
+									$processor->set_attribute( $key, $value );
+								}
+							}
+						}
+						$processor->remove_attribute( $attribute_name );
+					}
+
+					$value = $processor->get_attribute( $attribute_name );
+
+					if ( ! is_string( $value ) ) {
+						continue;
+					}
+
+					// Replace entire attributes if their content is exclusively a placeholder, e.g. `title="</%title>"`.
+					$full_match = null;
+					if ( preg_match( '~^</%([^>]+)>$~', $value, $full_match ) ) {
```

Comment:
`</%>` is a funky comment that could set the `""` (empty string) replacement. Was this a conscious decision to disallow empty string? I think in the HTML funky comment replacement it would handle this case.

I'm asking because I think the placeholder could be zero or more:

```suggestion
					if ( preg_match( '~^</%([^>]*)>$~', $value, $full_match ) ) {
```

The tag and attribute handling should behave consistently on this point, either empty funky comments are allowed or disallowed everywhere.

(There are several more places where this change would be necessary.)